service/cachemdw/cache.go

```diff
@@ -0,0 +1,128 @@
+package cachemdw
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/kava-labs/kava-proxy-service/clients/cache"
+	"github.com/kava-labs/kava-proxy-service/decode"
+	"github.com/kava-labs/kava-proxy-service/logging"
+)
+
+// ServiceCache is responsible for caching EVM requests and provides corresponding middleware
+// ServiceCache can work with any underlying storage which implements simple cache.Cache interface
+type ServiceCache struct {
+	cacheClient              cache.Cache
+	blockGetter              decode.EVMBlockGetter
+	cacheTTL                 time.Duration
+	decodedRequestContextKey any
+	// chainID is used as prefix for any key in the cache
+	chainID string
+
+	*logging.ServiceLogger
+}
+
+func NewServiceCache(
+	cacheClient cache.Cache,
+	blockGetter decode.EVMBlockGetter,
+	cacheTTL time.Duration,
+	decodedRequestContextKey any,
+	chainID string,
+	logger *logging.ServiceLogger,
+) *ServiceCache {
+	return &ServiceCache{
+		cacheClient:              cacheClient,
+		blockGetter:              blockGetter,
+		cacheTTL:                 cacheTTL,
+		decodedRequestContextKey: decodedRequestContextKey,
+		chainID:                  chainID,
+		ServiceLogger:            logger,
+	}
+}
+
+// IsCacheable checks if EVM request is cacheable.
+// In current implementation we consider request is cacheable if it has specific block height
+func IsCacheable(
+	ctx context.Context,
+	blockGetter decode.EVMBlockGetter,
+	logger *logging.ServiceLogger,
+	req *decode.EVMRPCRequestEnvelope,
+) bool {
+	blockNumber, err := req.ExtractBlockNumberFromEVMRPCRequest(ctx, blockGetter)
```

Comment:
if the request contains a block hash, this method makes a request to the chain to query block number, so this shouldn't be used in the critical path of responding.

the code i just merged broke out some utility methods from the original version you can use to avoid the chain call.

i'd recommend re-writing this `IsCacheable` method to use the following:
* [`MethodRequiresNoHistory`](https://github.com/Kava-Labs/kava-proxy-service/blob/main/decode/evm_rpc.go#L113) - requests that will always be cacheable
* [`MethodHasBlockHashParam`](https://github.com/Kava-Labs/kava-proxy-service/blob/main/decode/evm_rpc.go#L76) - is cacheable as long as there is a response 
  * note: don't cache results for hashes that don't exist, because someday the might
* [`MethodHasBlockNumberParam`](https://github.com/Kava-Labs/kava-proxy-service/blob/main/decode/evm_rpc.go#L53) - if this is true, then use [`ParseBlockNumberFromParams`](https://github.com/Kava-Labs/kava-proxy-service/blob/main/decode/evm_rpc.go#L264) to get the actual block number and return `blockNumber > 0 || blockNumber == decode.BlockTagToNumberCodec[decode.BlockTagEarliest]`
  * note that if `blockNumber > 0`, you again should only cache if the result is non-null. don't cache if the block is not found, because someday it will exist

Reply:
@evgeniy-scherbina re-emphasizing / agreeing with 

```
so this shouldn't be used in the critical path of responding.
```

this always need to happen, regardless of whether redis is down, or the caching middleware has bugs. The service needs to practice graceful degradation (as outlined here https://github.com/Kava-Labs/kava-proxy-service/blob/main/architecture/ARCHITECTURE.md?plain=1#L35 as job #1 for the proxy service)

Reply:
sorry for the ambiguity. `this` being `ExtractBlockNumberFromEVMRPCRequest`. it makes evm requests to the chain. we should remove them in favor of pure checks on whether the thing can be cached or not

ie we don't need to wait for an evm call to check what height a block hash is for to know that we can cache a request that includes a block hash

Reply:
how it makes evm calls:
`ExtractBlockNumberFromEVMRPCRequest` > `lookupBlockNumberFromHashParam`

https://github.com/Kava-Labs/kava-proxy-service/blob/37a680af8d28126ca06de6c887b2356b3e54c3f1/decode/evm_rpc.go#L252-L258

Reply:
good catch

fixed here: https://github.com/Kava-Labs/kava-proxy-service/pull/39/commits/07b4347682ead560ca4045d0f5c057890be4d533

implementation: https://github.com/Kava-Labs/kava-proxy-service/blob/yevhenii/cache/service/cachemdw/cache.go#L51

Reply:
Probably I have to also extend e2e-tests

As of now I only have tests for this group: `MethodHasBlockNumberParam`

But I have to add e2e-tests for this groups as well:
- `MethodHasBlockHashParam`
- `MethodRequiresNoHistory`